src/spi.rs

```diff
@@ -1298,9 +1638,124 @@ fn polling_transmit(
         length: (transaction_length * 8) as _,
         rxlength: (rx_length * 8) as _,
         ..Default::default()
-    };
+    }
+}
+
+fn spi_transmit(
+    handle: spi_device_handle_t,
+    transaction: &mut spi_transaction_t,
+    polling: bool,
+    _keep_cs_active: bool,
+) -> Result<(), EspError> {
+    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
+    // The CS pin should be kept active through transactions.
+    #[cfg(not(esp_idf_version = "4.3"))]
+    if _keep_cs_active {
+        transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
+    }
+
+    if polling {
+        esp!(unsafe { spi_device_polling_transmit(handle, transaction as *mut _) })
+    } else {
+        esp!(unsafe { spi_device_transmit(handle, transaction as *mut _) })
+    }
+}
+
+async fn spi_transmit_async(
+    handle: spi_device_handle_t,
+    transaction: &mut spi_transaction_t,
+    _keep_cs_active: bool,
+) -> Result<(), EspError> {
+    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
+    // The CS pin should be kept active through transactions.
+    #[cfg(not(esp_idf_version = "4.3"))]
+    if _keep_cs_active {
+        transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
+    }
+
+    let notification = Notification::new();
 
-    esp!(unsafe { spi_device_polling_transmit(handle, &mut transaction as *mut _) })
+    transaction.user = ptr::addr_of!(notification) as *mut _;
+
+    loop {
+        match esp!(unsafe {
+            spi_device_queue_trans(handle, transaction as *mut _, delay::NON_BLOCK)
+        }) {
+            Ok(_) => break,
+            Err(e) if e.code() != ESP_ERR_TIMEOUT => return Err(e),
+            _ => notification.wait().await,
+        }
+    }
```

Comment:
I don't agree it will deadlock and I don't understand why you believe it would deadlock given that it awaits the very same notification object that gets notified on any transaction initiation or completion. Also, I don't think it is possible for `spi_device_queue_trans` to actually return `ESP_ERR_TIMEOUT` given that we actually don't do transaction queueing ATM, but still - the current code should not hurt either.

Reply:
> I don't understand why you believe it would deadlock

If `spi_device_queue_trans` returns `ESP_ERR_TIMEOUT`, that means the transaction wasn't queued or processed at all, which means nothing will call `notification.notify()`, and `notification.wait()` will wait forever (or until cancellation).

Reply:
Let's put this aside for a moment; you **do** agree that - given how the code is arranged - returning `ESP_ERR_TIMEOUT` should never happen?

Reply:
Yes I agree `ESP_ERR_TIMEOUT` should never happen.

Reply:
OK. I've reworked the code to be in your explicit style. (But with an assertion on `ESP_ERR_TIMEOUT` never happening.)

And I agree - given that we only have `post_cb` and no `pre_cb` (just double checked that) we should be explicit.